aws_crt_common_ws/aws-crt-c-flags/src/lib.rs

```diff
@@ -0,0 +1,469 @@
+    }
+
+    /// Attempts to compile, `to_compile` and returns a result on whether or not it succeeded.
+    /// This is useful for testing compiler capabilities before including a file or flag in your build.
+    ///
+    /// # Arguments
+    ///
+    /// * `to_compile` - C code to attempt compilation of.
+    ///
+    /// # Examples
+    /// ```should_panic
+    /// use aws_crt_c_flags::{CRTModuleBuildInfo};
+    /// let mut build_info = CRTModuleBuildInfo::new("aws_crt_common_sys");
+    /// build_info.try_compile("int main() { return 0; }").expect("This should have compiled");
+    /// ```
+    pub fn try_compile(&self, to_compile: &str) -> core::result::Result<(), cc::Error> {
+        // try_compile prints linker stuff. We don't want that since this is just for testing the
+        // compiler capabilities. Suppress it for this scope.
+        let _suppress_stdout_cause_the_build_prints_linker_nonsense = Gag::stdout().unwrap();
+        let mut test_build = Build::new();
+        let output_location = format!(
+            "{}/compiler_checks",
+            env::var_os("OUT_DIR").unwrap().to_str().unwrap()
+        );
+        test_build.out_dir(&output_location);
+        fs::create_dir_all(&output_location).expect("creation of try compile directory failed");
+        let target_location = format!(
+            "{}/compiler_checks/check.c",
+            env::var_os("OUT_DIR").unwrap().to_str().unwrap()
+        );
+        fs::write(Path::new(&target_location.as_str()), to_compile).expect("File write failed");
+        test_build.file(&target_location);
+        let res = test_build.try_compile("test");
+        fs::remove_dir_all(&output_location).expect("Cleanup of try compile step failed!");
+        res
+    }
+
+    fn load_to_build(&mut self) {
+        // add default warning stuff.
+        if self.build_toolchain.get_compiler().is_like_msvc() {
+            self.add_private_cflag("/W4")
+                .add_private_cflag("/WX")
+                .add_private_cflag("/MP");
+            // relaxes some implicit memory barriers that MSVC normally applies for volatile accesses
+            self.add_private_cflag("/volatile:iso");
+            // disable non-constant initializer warning, it's not non-standard, just for Microsoft
+            self.add_private_cflag("/wd4204");
+            // disable passing the address of a local warning. Again, not non-standard, just for Microsoft
+            self.add_private_cflag("/wd4221");
+        } else {
+            self.add_private_cflag("-Wall")
+                .add_private_cflag("-Werror")
+                .add_private_cflag("-Wstrict-prototypes")
+                .add_private_cflag("-fno-omit-frame-pointer")
+                .add_private_cflag("-Wextra")
+                .add_private_cflag("-pedantic")
+                .add_private_cflag("-Wno-long-long")
+                .add_private_cflag("-fPIC");
+        }
+
+        if self.build_toolchain.is_flag_supported("-Wgnu").is_ok() {
```

Comment:
Don't have to do this now, but I feel like we should be able to encapsulate these checks in their own individual flag structs that implement a `CFlag` (or whatever) trait that encapsulates the check and the resultant flag per platform, and slap the flags on the `CRTModuleBuildInfo`

Reply:
Like as a consumer/user of this, I just want to opt-in to `WarningsAsErrors` or `PositionIndependentCode` or `NoGNUZeroVariadicMacroArguments`, etc.

```rust
let module = CRTModuleBuildInfo::new()
    .with_cflags(WarningsAsErrors)
    .with_cflags(PositionIndependentCode)
...
```